src/platform_impl/web/web_sys/mod.rs

```diff
@@ -60,35 +60,101 @@ impl WindowExtWebSys for Window {
+    let rect = raw.get_bounding_client_rect();
+    let style = window
+        .get_computed_style(raw)
+        .unwrap()
+        .expect("`getComputedStyle` returned `None`");
+
+    let display_none = style.get_property_value("display").unwrap() == "none";
+
+    let prop = |name| -> f64 {
+        let value = style.get_property_value(name).unwrap();
+        if display_none && name.starts_with("padding") {
+            // When `display` is `none`, the value returned for padding isn't
+            // guaranteed to be in `px` (it's left as a percentage if the
+            // property is specified as such, when normally it's resolved to
+            // `px`).
+            // Note: that's also true when `display` is `contents`, but for
+            // `<canvas>` that gets resolved to `display: none` and can never
+            // happen.
+            // So, return 0, since getting the size right isn't particularly
+            // important for an invisible element.
+            return 0.0;
+        }
+        // Remove the `px` from the end of the value and parse it.
+        value
+            .strip_suffix("px")
+            .expect("border and padding should always be in units of `px`")
+            .parse()
+            .unwrap()
+    };
+
+    Some(LogicalSize {
+        width: rect.width()
+            - prop("border-left-width")
+            - prop("border-right-width")
+            - prop("padding-left")
+            - prop("padding-right"),
+        height: rect.height()
+            - prop("border-top-width")
+            - prop("border-bottom-width")
+            - prop("padding-top")
+            - prop("padding-bottom"),
+    })
+}
+
+pub fn set_inner_size(raw: &HtmlCanvasElement, size: Size) {
     let scale_factor = scale_factor();
 
-    let physical_size = size.to_physical::<u32>(scale_factor);
-    let logical_size = size.to_logical::<f64>(scale_factor);
+    let mut logical_size = size.to_logical::<f64>(scale_factor);
+
+    if cfg!(not(feature = "css-size")) {
+        let physical_size = size.to_physical(scale_factor);
+        raw.set_width(physical_size.width);
+        raw.set_height(physical_size.height);
+    }
 
-    raw.set_width(physical_size.width);
-    raw.set_height(physical_size.height);
+    let window = web_sys::window().unwrap();
+    let style = window
+        .get_computed_style(raw)
+        // This can't fail according to the spec; I don't know why web-sys marks it as throwing and having an optional result.
+        .expect("`getComputedStyle` failed")
+        .expect("`getComputedStyle` returned `None`");
+
+    // This also can't fail according to the spec.
+    if style.get_property_value("box-sizing").unwrap() == "border-box" {
+        let prop = |name| -> f64 {
+            let value = style.get_property_value(name).unwrap();
+            // Cut off the last two characters to remove the `px` from the end.
+            value[..value.len() - 2].parse().unwrap()
+        };
```

Comment:
Nit: Deduplicate this helper closure into a function (since it's the same as above).